src/zepben/evolve/model/busbranch/bus_branch.py

```diff
@@ -0,0 +1,343 @@
+#  Copyright 2020 Zeppelin Bend Pty Ltd
+#
+#  This Source Code Form is subject to the terms of the Mozilla Public
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+from collections import defaultdict
+from enum import Enum
+from functools import reduce
+from typing import Set, Tuple, FrozenSet, Dict, Callable, Union, TypeVar, Optional
+
+from zepben.evolve import NetworkService, ConnectivityNode, Terminal, Switch, AcLineSegment, ConductingEquipment, \
+    PowerTransformer, \
+    EnergySource, EnergyConsumer, IdentifiedObject, ShuntCompensator, Junction, BusbarSection, Jumper, \
+    PerLengthSequenceImpedance, WireInfo
+
+__all__ = ["create_bus_branch_model", "CreationResult", "ErrorInfo", "ErrorType"]
+
+BBN = TypeVar('BBN')  # Bus-Branch Network
+TN = TypeVar('TN')  # Topological Node
+PTT = TypeVar('PTT')  # PowerTransformer Type
+TBT = TypeVar('TBT')  # Topological Branch Type
+
+
+class ErrorType(Enum):
+    invalid_number_of_terminals = "invalid_number_of_terminals"
+    unsupported_class = "unsupported_class"
+    missing_required_data = "missing_required_data"
+
+
+class ErrorInfo:
+    """
+    Error information for an identified object that failed a validation test when trying to generate a bus-branch model.
+    """
+
+    def __init__(self, error_type: ErrorType, io: IdentifiedObject):
+        self.error_type = error_type
+        self.io = io
+
+
+class CreationResult:
+    """
+    Represents the results of creating a bus-branch model from a network service (node-breaker-model)
+    """
+
+    def __init__(self, bus_branch_model: Optional[BBN] = None, errors: Dict[ErrorType, Set[ErrorInfo]] = defaultdict(set)):
+        self.bus_branch_model = bus_branch_model,
+        self.errors = errors
+
+    @property
+    def succeed(self) -> bool:
+        """True if no errors were found while trying to create the bus-branch model, False otherwise."""
+        return len(self.errors) == 0
+
+    def _add_error(self, error_type: ErrorType, io: IdentifiedObject):
+        self.errors[error_type].add(ErrorInfo(error_type, io))
+
+
+def create_bus_branch_model(
```

Comment:
so i think this is fine for now, but if we're expecting users to define all these things themselves in the future we should have an abstract class interface that they just extend/implement and provide to create_bus_branch_model. will enable us to have clearer documentation and a much simpler method signature here.

We should really probably do that regardless eventually, but i'm happy for this to go in as a first pass.

Reply:
Sounds good